Add additionally-supported CSP directives not listed in `Rack::Protection::ContentSecurityPolicy`

davetron5000 opened the pull request:
# Problem

`Rack::Protection::ContentSecurityPolicy` itemizes out the directives it will include in the header, however this list is missing some standard directives, such as `script-src-elem`.

This means you cannot use this middleware if you want to set those directives in your CSP header.

# Solution

I was not sure what made the most sense.  At first, I thought to add the missing directives to the list inside the class.  But then this list would need to be maintained.

Allowing *any* option to be a CSP directive doesn't exactly work since options for the middleware that aren't directives could not be distinguishes from options that *are* directives.

So what I did was a somewhat naive implementation that adds a `custom_directives` option.

```ruby
use Rack::Protection::ContentSecurityPolicy, custom_directives: {
  script_src_elem: 'https://example.com',
  style_src_attr: 'https://example.com',
}
```

I've only made one test and the impl, no doc updates or refactoring or anything.  Does this seem useful/reasonable?  Or would it be better to just add the missing directives?

davetron5000 committed:
allow CSP directives not listed in class

git-hub-dash reviewed (commented):
@davetron5000

dentarg commented:
> I was not sure what made the most sense. At first, I thought to add the missing directives to the list inside the class. But then this list would need to be maintained.

I hear you on this but I think it would be good to add the missing directives and have the ambition to maintain this middleware. Hopefully the community can help each other out with this. I think it is good knowledge sharing over everyone maintaining their in list of "custom directives".

davetron5000 committed:
remove test for "additional" properties

davetron5000 committed:
Add support directives + attempt at maintenance

davetron5000 commented:
@dentarg Thanks for the feedback - I've changed this to remove the introduced concept and just have a list of supported directives.

That said, I did make the change more drastic in hopes of allowing this to be more maintainable:

* Changed formatting from word-wrapped list to a single-directive-per-line list
* Ordered directives based on the spec document
* Separated the deprecated directives and included a link to explanation of their deprecation
* Slight change in how the no-arg directives are modeled

Happy to do a different way of formatting.  re: the single-directive-per-line, I just could not read the original wrapped list, but maybe it's just me :)

Also happy to lose the links to deprecation explanations so that we can use the `%i[]` syntax for both lists.

dentarg commented:
I haven't had the time to look at this, but I just found out about this: https://github.com/github/secure_headers 

(...because I was [searching the rails repo](https://github.com/search?q=repo%3Arails%2Frails+Sec-Fetch-Site&type=issues) for `Sec-Fetch-Site` because I was reading https://github.com/FiloSottile/csrf/blob/cfb6fbb0fbe34b0b2919ac81fa6dfe4678927dc5/csrf.go because I saw [this](https://ruby.social/@filippo@abyssdomain.expert/114519587678235700))

davetron5000 commented:
That gem looks useful, though I'm already seeing myself making doc PRs for it as I cannot for the life of me figure out how you would use it and/or how it becomes a Rack middleware without going into the code.

That all said - it does provide a much more richly-typed configuration for the available headers and their possible values.

rkh reviewed (approved):
Changes look good imo. Also, I agree we should keep an updated list. Would be a great opportunity to add an additional test or two.
